### PR TITLE
GitHub #635 standardise logging

### PR DESCRIPTION
**Pull request by ThrawnCA**
@KengoTODA @mebigfatguy @henrik242 @iloveeclipse How does this conversion look? I'll do similarly for other classes if we're happy with it.

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java [hunk 1]
```diff
@@ -20,28 +20,34 @@
...
+import java.util.logging.Logger;
 
 import javax.annotation.CheckForNull;
 
 /**
  * Version number and release date information.
  */
 public class Version {
+
+    private static final Logger LOGGER = Logger.getLogger(MethodHandles.lookup().lookupClass().getName());
```
Comment (iloveeclipse) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: I personally prefer just LOG.

Comment (iloveeclipse) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: Why not just Version.class.getName()?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java [hunk 2]
```diff
@@ -20,28 +20,34 @@
...
+
     /**
      * SpotBugs website.
      */
     public static final String WEBSITE = "https://spotbugs.github.io/";
 
-    public final static String VERSION_STRING;
+    public static final String VERSION_STRING;
```
Comment (iloveeclipse) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: I would avoid mixing unrelated changes, here and below too. even if this is due the style checks. PR's should not have side effects.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java [hunk 3]
```diff
@@ -99,22 +105,22 @@ public static void main(String[] argv) throws InterruptedException {
         String arg = argv[0];
 
         if ("-release".equals(arg)) {
-            System.out.println(VERSION_STRING);
+            LOGGER.info(VERSION_STRING);
```
Comment (iloveeclipse) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: I've never used JUL API: is the INFO level enabled by default?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java [hunk 4]
```diff
@@ -90,7 +96,7 @@ public static void registerApplication(String name, String version) {
         return applicationVersion;
     }
 
-    public static void main(String[] argv) throws InterruptedException {
+    public static void main(String[] argv) {
```
Comment (iloveeclipse) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: I'm not sure if main() the best place to introduce logging, as usually main() is called from the CLI and the output is (usually) expected to go to the STD OUT/ERR. Just assuming someone manages to redirect Java logging to a file, executing it will just don't do anything visible.

For the rest of the source code logging makes sense of course.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java [hunk 5]
```diff
@@ -99,22 +105,22 @@ public static void main(String[] argv) throws InterruptedException {
...
         if ("-release".equals(arg)) {
-            System.out.println(VERSION_STRING);
+            LOGGER.info(VERSION_STRING);
         } else if ("-plugins".equals(arg)) {
             DetectorFactoryCollection.instance();
             for(Plugin p : Plugin.getAllPlugins()) {
-                System.out.println("Plugin: " + p.getPluginId());
-                System.out.println("  description: " + p.getShortDescription());
-                System.out.println("     provider: " + p.getProvider());
+                LOGGER.log(Level.INFO, "Plugin: {0}", p.getPluginId());
```
Comment (iloveeclipse) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: I would use  `static import Level.*`, to be able to write it like:
`LOG.log(INFO, "Plugin: {0}", p.getPluginId());`

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 1 above)
Comment (ThrawnCA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: LOG or LOGGER works. But the class name is Logger, so...eh.

The reason for using `MethodHandles` is so that this line can be copy-pasted to any other class without modification.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 2 above)
Comment (ThrawnCA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: It's to keep SonarQube happy, yeah. Unrelated, but pretty trivial, so...

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 4 above)
Comment (ThrawnCA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: If logging is redirected to a file, then I would assume that that's the desired behavior, so I would hesitate to basically override it by forcing STDOUT here.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 3 above)
Comment (ThrawnCA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: It's configurable in `jre/lib/logging.properties`, but yes, the default level in that file is INFO.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 5 above)
Comment (ThrawnCA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: I guess it's a matter of style. Is there any risk of `Level` constants clashing with anything else we might want to static-import (I'm particularly thinking of ERROR)?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 2 above)
Comment (ThrawnCA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: @iloveeclipse Which approach would you prefer? Second pull request on the same branch after the first one is merged? Separate Github issue just saying "cleanup for issue X"?

**Comment by KengoTODA** [comment]
If this is for `master` branch, I think it's acceptable to introduce SLF4J. It's not so big, and really stable logging facade nowadays (I'm personally not sure Log4J 2's stability and usability).

I think JUL is just a temporal workaround to make logging logic easy to read without additional dependency.

**Comment by ThrawnCA** [comment]
@iloveeclipse @KengoTODA Any consensus on JUL vs SLF4J for a first pass? And does this pull request look OK as a starting point? I want to migrate other classes that are using STDOUT and STDERR, but let's first settle on a direction.

**Comment by iloveeclipse** [comment]
I'm OK with either one, but  I'm not a logging fetishist, it should just work.

**Comment by ThrawnCA** [comment]
OK, I'll call that a vote for SLF4J plus an abstention. Will update the pull request to use SLF4J.

**Comment by KengoTODA** [comment]
Thanks @ThrawnCA, I also vote for SLF4J as n author of [findbugs-slf4j](https://github.com/KengoTODA/findbugs-slf4j).

**Comment by mebigfatguy** [comment]
slf4j +1

**Comment by ThrawnCA** [comment]
Attempting to convert to SLF4J results in a problem like https://github.com/spotbugs/spotbugs-gradle-plugin/issues/11

**Comment by ThrawnCA** [comment]
Looks like the needs-work review from @iloveeclipse takes priority over the approval. Do I need to make further changes, or are you OK with it as-is?

File: spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java (hunk 2 above)
Comment (iloveeclipse) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/Version.java]: Cleanup after the main fix. This way one can always do cherry picking without side effects. So yes, please next time try to not mix different changes in one commit.
